src/vm/analysis/trait_checker/pre_type_check.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::{HashMap, HashSet};
+
+use vm::analysis::types::{ContractAnalysis, AnalysisPass};
+use vm::analysis::AnalysisDatabase;
+use vm::analysis::errors::{CheckResult, CheckError, CheckErrors};
+use vm::representations::{SymbolicExpression, ClarityName};
+use vm::representations::SymbolicExpressionType::{AtomValue, Atom, List, LiteralValue, TraitReference, Field};
+use vm::types::{Value, TraitIdentifier};
+use vm::functions::NativeFunctions;
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+
+pub struct PreTypeCheckingTraitChecker {
+}
+
+impl AnalysisPass for PreTypeCheckingTraitChecker {
+
+    fn run_pass(contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let mut command = PreTypeCheckingTraitChecker::new();
+        command.run(contract_analysis, analysis_db)?;
+        Ok(())
+    }
+}
+
+impl PreTypeCheckingTraitChecker {
+
+    fn new() -> Self {
+        Self {
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let exprs = contract_analysis.expressions[..].to_vec();
+        let mut trait_usages = self.find_trait_usages(&exprs)?;
+        
+        // Presence of orphaned traits should throw
+        for (orphan, expr) in trait_usages.orphan_trait_references {
+            let mut err = CheckError::new(CheckErrors::TraitReferenceUnknown(orphan.to_string()));
+            err.set_expression(&expr);
+            return Err(err.into());
+        }
+
+        // Ensure that imported traits exists
+        for (trait_name, trait_expr) in trait_usages.imported_traits.drain() {
+            let imported_trait_args = trait_expr.match_list().ok_or(CheckErrors::ImportTraitBadSignature)?;
+            if imported_trait_args.len() != 3 {
+                return Err(CheckErrors::ImportTraitBadSignature.into())
+            }
+            let trait_identifier = match &imported_trait_args[2].expr {
+                Field(field) => field,
+                _ => return Err(CheckErrors::ImportTraitBadSignature.into()),
+            };
+
+            let existing_trait = analysis_db.get_defined_trait(
+                &trait_identifier.contract_identifier, 
+                &trait_identifier.name)?;
+            existing_trait.ok_or(CheckError::new(CheckErrors::TraitReferenceUnknown(trait_name.to_string())))?;
+            
+            // Check for eventual collisions
+            if contract_analysis.referenced_traits.contains_key(&trait_name) {
+                return Err(CheckErrors::NameAlreadyUsed(trait_name.to_string()).into())
```

Comment:
The early return here would cause multiple instances of this error (using the same trait name more than once) to get dropped.  Is that the intended behavior?

Reply:
Yes, our current structure doesn't allow us to carry and show multiple diagnostics at once.